### PR TITLE
Add filters for customer email and order id for performance.

### LINKED ISSUE #41454: [Enhancement]: Enable HPOS, I got very slow queries in a more than 400,000 orders store
### Describe the solution you'd like

before enable HPOS, I search a order need 6 seconds, the slow query is about 3-5s, then afte i enable HPOS in test site, i got slow query recored more than 10 seconds, 
the mysql sql is 
```
SELECT wp_wc_orders.id
FROM wp_wc_orders
LEFT JOIN wp_woocommerce_order_items AS search_query_items
ON search_query_items.order_id = wp_wc_orders.id
WHERE 1=1
AND (wp_wc_orders.status IN ('wc-pending','wc-processing','wc-on-hold','wc-completed','wc-cancelled','wc-refunded','wc-failed','wc-partial-payment'))
AND (wp_wc_orders.type = 'shop_order')
AND ( ( `wp_wc_orders`.id = 672253
OR search_query_items.order_item_name LIKE '%672253%'
OR `wp_wc_orders`.id IN (
SELECT search_query_meta.order_id
FROM wp_wc_orders_meta as search_query_meta
WHERE search_query_meta.meta_key IN ( '_billing_address_index','_shipping_address_index' )
AND search_query_meta.meta_value LIKE '%672253%'
GROUP BY search_query_meta.order_id ) ) )
GROUP BY wp_wc_orders.id
ORDER BY wp_wc_orders.date_created_gmt DESC
LIMIT 0, 20
```
here is Query Monitor result,
<a href="https://ibb.co/FK9GmqP">slow-query</a>
So can this mysql statement still be optimized?

-----------------
update

I try to test some different queries

first, original mysql statement
```
SELECT
	wp_wc_orders.id 
FROM
	wp_wc_orders
	LEFT JOIN wp_woocommerce_order_items AS search_query_items ON search_query_items.order_id = wp_wc_orders.id 
WHERE
	1 = 1 
	AND (
	wp_wc_orders.STATUS IN ( 'wc-pending', 'wc-processing', 'wc-on-hold', 'wc-completed', 'wc-cancelled', 'wc-refunded', 'wc-failed', 'wc-partial-payment' )) 
	AND ( wp_wc_orders.type = 'shop_order' ) 
	AND (
		(
			`wp_wc_orders`.id = 672253 
			OR search_query_items.order_item_name LIKE '%672253%' 
			OR `wp_wc_orders`.id IN (
			SELECT
				search_query_meta.order_id 
			FROM
				wp_wc_orders_meta AS search_query_meta 
			WHERE
				search_query_meta.meta_key IN ( '_billing_address_index', '_shipping_address_index' ) 
				AND search_query_meta.meta_value LIKE '%672253%' 
			GROUP BY
				search_query_meta.order_id 
			) 
		) 
	) 
GROUP BY
	wp_wc_orders.id 
ORDER BY
	wp_wc_orders.date_created_gmt DESC 
	LIMIT 0,
	20
> OK
> Time: 14.046s
```

second, delete the parts related to wp_woocommerce_order_items table
```
SELECT
    wp_wc_orders.id
FROM
    wp_wc_orders
WHERE
    1 = 1
    AND wp_wc_orders.status IN ('wc-pending', 'wc-processing', 'wc-on-hold', 'wc-completed', 'wc-cancelled', 'wc-refunded', 'wc-failed', 'wc-partial-payment')
    AND wp_wc_orders.type = 'shop_order'
    AND (
        wp_wc_orders.id = 672253
        OR wp_wc_orders.id IN (
            SELECT
                search_query_meta.order_id
            FROM
                wp_wc_orders_meta AS search_query_meta
            WHERE
                search_query_meta.meta_key IN ('_billing_address_index', '_shipping_address_index')
                AND search_query_meta.meta_value LIKE '%672253%'
            GROUP BY
                search_query_meta.order_id
        )
    )
GROUP BY
    wp_wc_orders.id
ORDER BY
    wp_wc_orders.date_created_gmt DESC
LIMIT 0, 20
> OK
> Time: 31.381s
```
3th, delete the parts related to wp_woocommerce_order_items and wp_wc_orders_meta table
```
SELECT
    wp_wc_orders.id
FROM
    wp_wc_orders
WHERE
    wp_wc_orders.status IN ('wc-pending', 'wc-processing', 'wc-on-hold', 'wc-completed', 'wc-cancelled', 'wc-refunded', 'wc-failed', 'wc-partial-payment')
    AND wp_wc_orders.type = 'shop_order'
    AND wp_wc_orders.id = 672253
GROUP BY
    wp_wc_orders.id
ORDER BY
    wp_wc_orders.date_created_gmt DESC
LIMIT 0, 20
> OK
> Time: 0.174s
```
Woocommerce's default order search function is very convenient for a small number of orders. It can search for any content, including but not limited to order number, email, name, address, etc. 
But for customers with a large number of orders, this A combined search will make the search very slow. Even if the server configuration is improved and the database is optimized, the degree of improvement is limited. Both the previous database table and the current HPOS have the same problem. 
So can you add some new functions (such as selecting elements) to the search place? We can choose to search all data, or specified data fields, such as order id, email, etc.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vedanshujain**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #41454 

This PR adds two more options to the search filter. Most likely we would need another PR to add filter so that folks can add their own custom search boxes, but before we do that I want to wait and see how these changes are received. 

We added an order ID filter and customer email filter, which as named, will limit the search to only those fields. They also remove wildcard, and both are indexed fields with HPOS (ID was already an indexed field with posts as well, but now billing email is also indexed).

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. With HPOS authoritative, create a few orders (at least 3). In at least one of the orders, use a unique email, make sure to note that email.
2. Go to admin orders on `wp-admin/admin.php?page=wc-orders`
3. Select `Customer Email` as the search option, and enter the unique email used in step 1 to search. Check that the order with that email shows up.
4. Similarly, add any order ID in the search box, select order ID in the search dropdown and search. Check that order with that ID shows up.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php [hunk 1]
```diff
@@ -158,6 +163,21 @@ private function generate_where_for_search_filter( string $search_filter ) : str
 
 		$order_table = $this->query->get_table_name( 'orders' );
 
+		if ( 'customer_email' === $search_filter ) {
+
+			return $wpdb->prepare(
```
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php]:
```suggestion
		if ( 'customer_email' === $search_filter ) {
			return $wpdb->prepare(
```

**Review by coreymckrill — COMMENTED** [review]
👍 Works as described testing locally. A couple thoughts:

> We added an order ID filter and customer email filter, which as named, will limit the search to only those fields. They also remove wildcard, and both are indexed fields with HPOS

Not having wildcard on the order ID search makes sense to me. However, for an email search, I would always expect a partial string match to work, and it doesn't in this case. Having it only match on exact email address may be faster, but I think it's not very useful to the user.

**Comment by vedanshujain** [comment]
> However, for an email search, I would always expect a partial string match to work, and it doesn't in this case. Having it only match on exact email address may be faster, but I think it's not very useful to the user.

I have added a trailing wildcard, that way it can still use the index. For a full match, the `customer` or `all` filter can still be used. wdyt about shipping like this and if we get feedback that trailing wildcard is also not good enough we will add a leading one too.

**Review by coreymckrill — APPROVED** [review]
👍 Yep, I think a trailing wildcard is good enough, at least until we get feedback to the contrary.

**Comment by robertstaddon** [comment]
Would it be possible to add a WordPress hook so that so that a site with 150k+ orders can change the default order search from the slow "All" option to one of the faster order search filters?

Also, which one of the available search filters is closest to the way the pre-HPOS search worked? That would be really helpful to set as a fast default order search filter.

**Comment by coreymckrill** [comment]
@robertstaddon I think that's worth considering. Could you open a new enhancement request issue for this?

**Comment by robertstaddon** [comment]
Thank you for the suggestion, @coreymckrill , I've created the enhancement request here: https://github.com/woocommerce/woocommerce/issues/45721

**Pull request merged**